tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/SelectSeqAsFold.scala

```diff
@@ -0,0 +1,87 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir.oper.{ApalacheOper, TlaActionOper, TlaControlOper, TlaFunOper, TlaOper, TlaSeqOper}
+import at.forsyte.apalache.tla.lir.{
+  BoolT1, LetInEx, NameEx, OperEx, OperParam, OperT1, SeqT1, TlaEx, TlaOperDecl, Typed, TypingException
+}
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+
+/**
+ * Replaces instances of SelectSeq, using the following definition:
+ *
+ * SelectSeq(seq, Test(_)) == LET CondAppend(s,e) == IF Test(e) THEN Append(s, e) ELSE s
+ *                            IN FoldSeq( CondAppend, <<>>, seq )
+ */
+class SelectSeqAsFold(nameGenerator: UniqueNameGenerator, tracker: TransformationTracker) extends TlaExTransformation {
+
+  override def apply(expr: TlaEx): TlaEx = {
+    transform(expr)
+  }
+
+  def transform: TlaExTransformation = tracker.trackEx {
+    case OperEx(TlaSeqOper.selectseq, seqEx, testNameEx @ NameEx(operName)) =>
+      // Sanity check on types
+      val seqElemType = seqEx.typeTag match {
+        case Typed(SeqT1(a)) => a
+        case badType         => throw new TypingException(s"Argument $seqEx expected to have type Seq(a), found $badType.")
```

Comment:
Is type checking not done by this point in the rewriting?

Reply:
It should be, but I have to do a match to read the inner `a` value anyway, so the other case is just a safeguard.

Reply:
If it should be type checked, I'd suggest not repeating the checks here. That can lead to maintenance issues and weird divergence of behavior. I think it's better to just ensure we do the checking properly in the appropriate phase.

Reply:
Counterpoint: writing such sanity checks (which are cheap anyway, because I need to pattern match with the first case anyway) actually frees us from the implicit assumption and dependency, and it makes this pass movable in the pass order. Pass dependencies have historically been a big pain point.

Reply:
My concern is not for the computational cost, but for the maintenance. 

If we end up moving this above the type checking level, then, wouldn't it be incorrect for it to be doing it's own type checking? To my mind, this actually makes the transformation less portable and less general. (E.g., it cannot be used on untyped code.)

If we end up imposing an improved system for giving errors on type checking fails, and we've peppered these kinds of adhoc sanity checks through the code, we're going to have to track down every case do duplicate work. Or else maybe we miss them, and we end up with two different kinds of errors reported. Ditto regarding unnecessary re-work if we change the name or representation of the type.

I won't block on this, but I don't see how it's a value add, and do foresee allowing this kind of sanity check to lead to maintenance issues and needless code bloat.